clar2wasm/src/standard/standard.wat

```diff
@@ -178,182 +178,153 @@
         (return (local.get $diff_lo) (local.get $diff_hi))
     )
 
-    (func $mul-uint (type 1) (param $a_lo i64) (param $a_hi i64) (param $b_lo i64) (param $b_hi i64) (result i64 i64)
-        (local $a0 i32)
-        (local $a1 i32)
-        (local $a2 i32)
-        (local $a3 i32)
-        (local $b0 i32)
-        (local $b1 i32)
-        (local $b2 i32)
-        (local $b3 i32)
-        (local $product0 i64)
-        (local $product1 i64)
-        (local $product2 i64)
-        (local $product3 i64)
-        (local $carry i64)
-        (local $res_hi i64)
-        (local $res_lo i64)
-
-        ;; Shortcut if either a or b is zero
-        (if (i32.or
-                (i64.eqz (i64.or (local.get $a_hi) (local.get $a_lo)))
-                (i64.eqz (i64.or (local.get $b_hi) (local.get $b_lo))))
-            (return (i64.const 0) (i64.const 0))
+    (func $mul-int128 (type 1) (param $a_lo i64) (param $a_hi i64) (param $b_lo i64) (param $b_hi i64) (result i64 i64)
+    ;; Adaptation of Hacker's Delight, chapter 8
+    ;; u1 <- $a_lo & 0xffffffff; v1 <- $b_lo & 0xffffffff
+    ;; u2 <- $a_lo >> 32; v2 <- $b_lo >> 32
+    ;; t1 <- v1 * u1
+    ;; t2 <- (u2 * v1) + (t1 >> 32)
+    ;; t3 <- (u1 * v2) + (t2 & 0xffffffff)
+    ;; $res_lo <- (t2 << 32) | (t1 & 0xffffffff)
+    ;; $res_hi <- ($a_lo * b_hi) + ($a_hi * b_lo) + (v2 * u2) + (t2 >> 32) + (t3 >> 32)
+        (local $u2 i64) (local $v2 i64) (local $t1 i64) (local $t2 i64) (local $t3 i64)
+        (i64.or
+            (i64.shl
+                (local.tee $t3 (i64.add
+                    (i64.and
+                        (local.tee $t2 (i64.add 
+                            (i64.shr_u
+                                (local.tee $t1 (i64.mul
+                                    ;; $v2 contains u1 temporarily
+                                    (local.tee $v2 (i64.and (local.get $a_lo) (i64.const 0xffffffff))) 
+                                    ;; $u2 contains v1 temporarily
+                                    (local.tee $u2 (i64.and (local.get $b_lo) (i64.const 0xffffffff))) 
+                                ))
```

Comment:
(opinionated voice) Please put closing brackets on the same line.

I really think we should follow general lisp style in our wat and clarity code. See [this style guide](https://mumble.net/~campbell/scheme/style.txt)

Reply:
I agree that there should be a definitive style for how we write WAT. However, I don’t think this is the right PR for this. 
Since most of the current document is written with this current style, I believe we should keep it until we decide which actual style we want to adopt. 
It might be the scheme style, actually, that could be great if we could find something like _prettier_ which implements this style and that we can enforce in the CI.

Reply:
I haven't had luck trying a Clojure formatter. I think someone might need to complete a side-quest of implementing a WAT formatter if we want this.

Reply:
but also agree, let's figure that out in a PR of its own.